components/automate-deployment/pkg/converge/converger.go

```diff
@@ -127,23 +121,315 @@ func StartConverger(opts ...ConvergerOpt) Converger {
+//            ^       |
+//            |       |
+//   timeout 2|       |configureRequest
+//            |       |(err = ErrSelfReconfigurePending)
+//            |       |
+//            |       |
+//            |       v
+//         +--+--------------------+
+//         |                       | timeout 1
+//         | waitingForReconfigure +---------+
+//         |                       |         |
+//         |                       |<--------+
+//         +-----------------------+
+//
+
+// idleState is the default state of the converger. In this state we
+// are ready to accept any user-sendable requests. The idle state
+// never times out.
+type idleState struct{}
+
+func (t *idleState) ProcessMessage(c Compiler, m message) (state, error) {
+	switch m.(type) {
+	case *convergeRequest:
+		req := m.(*convergeRequest)
+		convergePlan, err := c.Compile(req.desiredState)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to compile")
+			return t, err
+		}
+		err = convergePlan.Execute(req.eventSink)
+		return t.handleError(err)
+	case *stopRequest:
+		req := m.(*stopRequest)
+		convergePlan, err := c.Compile(req.desiredState)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to compile")
+			return t, err
+		}
+		err = convergePlan.Execute(req.eventSink)
+		return t.handleError(err)
+	case *timeout:
+		logrus.Warn("Received Timeout when in idle state.")
+		return &idleState{}, nil
+	default:
+		return t, nil
+	}
+}
+
+func (t *idleState) handleError(err error) (state, error) {
+	switch err {
+	case nil:
+		return t, nil
+	case api.ErrRestartPending:
+		return &waitingForRestart{}, err
+	case api.ErrSelfReconfigurePending:
+		return &waitingForReconfigure{}, err
+	case api.ErrSelfUpgradePending:
+		// NOTE(ssd) 2019-04-19: A self-upgrade is
+		// essentially just a restart from hab-sup
+		return &waitingForRestart{}, err
+	default:
+		return t, err
+	}
+}
+
+func (t *idleState) Timeout() time.Duration { return time.Duration(0) }
+func (t *idleState) String() string         { return "idle" }
+
+// waitingForRestart is a converger state that we enter whenever we
+// expect to be restarted by an external signal.
+type waitingForRestart struct{}
+
+func (t *waitingForRestart) ProcessMessage(_ Compiler, m message) (state, error) {
+	switch m.(type) {
+	case *timeout:
+		logrus.Info("timed out waiting for restart, re-entering idle state")
+		return &idleState{}, nil
+	default:
+		logrus.Infof("ignoring message %v because we are waiting to be restarted", m)
+		return t, nil
+	}
+}
+
+func (t *waitingForRestart) Timeout() time.Duration { return 600 * time.Second }
+func (t *waitingForRestart) String() string         { return "waiting for restart" }
+
+// waitingForReconfigure is a converer state that we enter whenever we
+// expect to have received a reconfiguration request from the Habitat
+// supervisor.
+type waitingForReconfigure struct {
+	sentSelfHup bool
```

Comment:
Maybe rather than having this boolean, we can have two states: `waitingForReconfigure1` and `waitingForReconfigure2`.